grudge/dt_utils.py

```diff
@@ -271,31 +271,50 @@ def dt_geometric_factors(
     dd_face = DOFDesc("all_faces", dd.discretization_tag)
     face_discr = dcoll.discr_from_dd(dd_face)
 
-    # To get a single value for the total surface area of a cell, we
-    # take the sum over the averaged face areas on each face.
-    # NOTE: The face areas are the *same* at each face nodal location.
-    # This assumes there are the *same* number of face nodes on each face.
-    surface_areas = abs(
+    # Compute areas of each face
+    face_areas = abs(
         op.elementwise_integral(
             dcoll, dd_face, face_discr.zeros(actx) + 1.0
         )
     )
-    surface_areas = DOFArray(
-        actx,
-        data=tuple(
-            actx.einsum("fej->e",
-                        face_ae_i.reshape(
-                            vgrp.mesh_el_group.nfaces,
-                            vgrp.nelements,
-                            afgrp.nunit_dofs
-                        ),
-                        tagged=(FirstAxisIsElementsTag(),)) / afgrp.nunit_dofs
-
-            for vgrp, afgrp, face_ae_i in zip(volm_discr.groups,
-                                              face_discr.groups,
-                                              surface_areas)
+
+    if actx.supports_nonscalar_broadcasting:
+        # Compute total surface area of an element by summing over the
+        # individual face areas
+        surface_areas = DOFArray(
+            actx,
+            data=tuple(
+                actx.einsum(
+                    "fej->e",
+                    face_ae_i.reshape(
+                        vgrp.mesh_el_group.nfaces, vgrp.nelements, -1),
```

Comment:
```suggestion
                    "fe->e",
                    face_ae_i.reshape(
                        vgrp.mesh_el_group.nfaces, vgrp.nelements),
```

Reply:
@inducer I tried this suggestion but unfortunately, I am running into this `ValueError` when I run the unit tests in `test/test_dt_utils.py`: https://github.com/inducer/pytato/blob/main/pytato/array.py#L1059 --- I think the routine is still expecting 3 indices in the einsum signature

Reply:
Fair enough, thanks for trying.